File: python/global_matrix.py
import scipy.sparse

def create(N, c, f, BC, A0, A1, A2):
	A0 = 	[ A0[2,2], A0[2,3], A0[3,2], A0[3,3] ] + 
			(N-2)*[ i for a in A0 for i in a ] +
			[ A0[0,0], A0[0,1], A0[3,1], A0[1,1] ]
	
	A1 = 	[ A1[2,2], A1[2,3], A1[3,2], A1[3,3] ] + 
			(N-2)*[ i for a in A1 for i in a ] +
			[ A1[0,0], A1[0,1], A1[3,1], A1[1,1] ]
	
	A2 = 	[ A2[2,2], A2[2,3], A2[3,2], A2[3,3] ] + 
			(N-2)*[ i for a in A2 for i in a ] +
			[ A2[0,0], A2[0,1], A2[3,1], A2[1,1] ]
	
	A = [ c[0]*A0[i] + c[1]*A1[i] + c[2]*A2[i] for i in range(len(A0)) ]
	
	row = [ i  for j in range(2, 2*N+2, 2)  for i in range(j-2, j+2)  for k in range(4) ]
	column = [ i  for j in range(2, 2*N+2, 2)  for k in range(4)  for i in range(j-2, j+2) ]
	
	A = scipy.sparse.coo_matrix( (A, (row, column)), shape=(2*N-2, 2*N-2))
	A = A.tobsr()
	
	
	A0 = scipy.sparse.coo_matrix( (A0, (row, column)), shape=(2*N-2, 2*N-2) )
	A0 = A0.tobsr()
	f = np.array( [ f(x//2/N)[x%2] for x in range(2,2*N) ] )
	# S fv dx
	F = A0*f
	
	return (A, F)

File: matlab/global_matrix.m
function [A, F] = global_matrix(N, c, f, BC, A0, A1, A2)
	% make global matrix A s.t. A*U = F
	% remove A[0:2,:],  A[:,0:2],  A[-2:,:], A[:,-2:] to add to RHS later
	A0_ = 	[ [ A0(3,3), A0(3,4), A0(4,3), A0(4,4) ] ...
			repmat(A0'(:)',1,N-2) ...
			[ A0(1,1), A0(1,2), A0(2,1), A0(2,2) ] ];
	A1_ = 	[ [ A1(3,3), A1(3,4), A1(4,3), A1(4,4) ] ...
			repmat(A1'(:)',1,N-2) ...
			[ A1(1,1), A1(1,2), A1(2,1), A1(2,2) ] ];
	A2_ = 	[ [ A2(3,3), A2(3,4), A2(4,3), A2(4,4) ] ...
			repmat(A2'(:)',1,N-2) ...
			[ A2(1,1), A2(1,2), A2(2,1), A2(2,2) ] ];
	A = 	c(1)*A0_ + c(2)*A1_ + c(3)*A2_;
	
	row = zeros(size(A0_));
	n = 5;
	for j = 3:2:2*N-2
		for i = j-2:j+1
			row(n:n+3) = [i, i, i, i];
			n = n+4;
		end
	end
	row(1:4) = [1, 1, 2, 2];
	row(end-3:end) = [2*N-3, 2*N-3, 2*N-2, 2*N-2];
	
	col = zeros(size(A0_));
	n = 5;
	for j = 3:2:2*N-2
		for i = 1:4
			col(n:n+3) = j-2:j+1;
			n = n+4;
		end
	end
	col(1:4) = [1, 2, 1, 2];
	col(end-3:end) = [2*N-3, 2*N-2, 2*N-3, 2*N-2];
	
	A = sparse(row, col, A, 2*N-2, 2*N-2);
	
	% make F for A*U = F
	% move A[0:2,:],  A[:,0:2],  A[-2:,:], A[:,-2:] to add to RHS
	A_ = c(1)*A0 + c(2)*A1 + c(3)*A2;
	bc1 = BC(1)*A_(:,1) + BC(2)*A_(:,2);
	bc2 = BC(3)*A_(:,3) + BC(4)*A_(:,4);
	% create A0 to integrate f*v
	A0_ = 	repmat(A0'(:),1,N);
	
	row = zeros(size(A0_));
	n = 1;
	for j = 3:2:2*N+2
		for i = j-2:j+1
			row(n:n+3) = [i, i, i, i];
			n = n+4;
		end
	end
	
	col = zeros(size(A0_));
	n = 1;
	for j = 3:2:2*N+2
		for i = 1:4
			col(n:n+3) = j-2:j+1;
			n = n+4;
		end
	end
	A0 = sparse(row, col, A0_, 2*N+2, 2*N+2);
	F = zeros(2*N+2,1);
	i = 1;
	for x = 0:2*N+1
		F(i) = f(floor(x/2)/(N-1))(mod(x,2)+1);
		i = i+1;
	end
	% S fv dx
	F = A0*F;
	F(1:4) = F(1:4) - bc1;
	F(end-3:end) = F(end-3:end) - bc2;
	F = F(3:end-2);
end